src/Native/Unix/System.Native/pal_signal.c

```diff
@@ -0,0 +1,287 @@
+        // Since we overwrote the disposition, we are now responsible for reaping those processes.
+        pid_t pid;
+        do
+        {
+            int status;
+            while ((pid = waitpid(-1, &status, WNOHANG)) < 0 && errno == EINTR);
+        } while (pid > 0);
+    }
+    else if ((void*)origHandler->sa_sigaction == (void*)SIG_DFL)
+    {
+        // do nothing
+    }
+    else if (origHandler->sa_sigaction != NULL)
+    {
+        // TODO?: We are passing a NULL siginfo and context, do we need to try and do better?
+        origHandler->sa_sigaction(SIGCHLD, NULL, NULL);
+    }
+}
+
+// Entrypoint for the thread that handles signals where our handling
+// isn't signal-safe.  Those signal handlers write the signal to a pipe,
+// which this loop reads and processes.
+void* SignalHandlerLoop(void* arg)
+{
+    // Passed in argument is a ptr to the file descriptor
+    // for the read end of the pipe.
+    assert(arg != NULL);
+    int pipeFd = *(int*)arg;
+    free(arg);
+    assert(pipeFd >= 0);
+
+    // Continually read a signal code from the signal pipe and process it,
+    // until the pipe is closed.
+    while (true)
+    {
+        // Read the next signal, trying again if we were interrupted
+        uint8_t signalCode;
+        ssize_t bytesRead;
+        while ((bytesRead = read(pipeFd, &signalCode, 1)) < 0 && errno == EINTR);
+
+        if (bytesRead <= 0)
+        {
+            // Write end of pipe was closed or another error occurred.
+            // Regardless, no more data is available, so we close the read
+            // end of the pipe and exit.
+            close(pipeFd);
+            return NULL;
+        }
+
+        if (signalCode == SIGQUIT || signalCode == SIGINT)
+        {
+            // We're now handling SIGQUIT and SIGINT. Invoke the callback, if we have one.
+            CtrlCallback callback = g_ctrlCallback;
+            int rv = callback != NULL ? callback(signalCode == SIGQUIT ? Break : Interrupt) : 0;
+            if (rv == 0) // callback removed or was invoked and didn't handle the signal
+            {
+                // In general, we now want to remove our handler and reissue the signal to
+                // be picked up by the previously registered handler.  In the most common case,
+                // this will be the default handler, causing the process to be torn down.
+                // It could also be a custom handler registered by other code before us.
+                UninitializeConsole();
+                sigaction(signalCode, OrigActionFor(signalCode), NULL);
+                kill(getpid(), signalCode);
+            }
+        }
+        else if (signalCode == SIGCHLD)
+        {
+            SigChldCallback callback = g_sigChldCallback;
+            if (callback != NULL)
+            {
+                callback();
+            }
+            else
+            {
+                SystemNative_ResumeSigChld();
```

Comment:
In what situation does this occur?  Is it only when no process has ever been started with Process.Start yet in this process, or is it possible for g_sigChldCallback to go from non-null to null?  I'm wondering about the waitpids this is doing and whether it'll interfere with a Process later getting the results of a process when code Waits on it and accesses its exit information.

Reply:
This occurs when signal handling was setup for the Console and no Process.Starts have occurred.
I'll add a comment.